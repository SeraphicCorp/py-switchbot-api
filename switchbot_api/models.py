"""constant for  SwitchBot API."""

from __future__ import annotations

from enum import Enum, StrEnum


class PowerState(Enum):
    """Power state."""

    ON = "on"
    OFF = "off"


class BatteryCirculatorFanMode(StrEnum):
    """Fan mode types [Battery Circulator Fan] API."""

    DIRECT = "direct"
    NATURAL = "natural"
    SLEEP = "sleep"
    BABY = "baby"


class VacuumFanSpeed(StrEnum):
    """Fan options for VacuumCommands supported devices."""

    VACUUM_FAN_SPEED_QUIET = "0"
    VACUUM_FAN_SPEED_STANDARD = "1"
    VACUUM_FAN_SPEED_STRONG = "2"
    VACUUM_FAN_SPEED_MAX = "3"


class VacuumFanSpeedV2(StrEnum):
    """Fan options for VacuumV2Commands & VacuumV3Commands supported devices."""

    VACUUM_FAN_SPEED_QUIET = "1"
    VACUUM_FAN_SPEED_STANDARD = "2"
    VACUUM_FAN_SPEED_STRONG = "3"
    VACUUM_FAN_SPEED_MAX = "4"


class VacuumCleanMode(StrEnum):
    """Clean mode for Vacuum."""

    SWEEP = "sweep"
    MOP = "mop"
    SWEEP_MOP = "sweep_mop"
