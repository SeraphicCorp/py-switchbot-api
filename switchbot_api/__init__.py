"""Tools to query the SwitchBot API."""
import base64
from dataclasses import dataclass
from enum import Enum
import hashlib
import hmac
import logging
import time
from typing import TypeVar
import uuid

from aiohttp import ClientSession

_API_HOST = "https://api.switch-bot.com"

_LOGGER = logging.getLogger(__name__)
NON_OBSERVED_REMOTE_TYPES = ["Others"]


class CannotConnect(Exception):
    """Cannot connect to the SwitchBot API."""


class InvalidAuth(Exception):
    """Invalid auth for the SwitchBot API."""


@dataclass
class Device:
    """Device."""

    device_id: str
    device_name: str
    device_type: str
    hub_device_id: str

    def __init__(self, **kwargs) -> None:
        """Initialize."""
        self.device_id = kwargs["deviceId"]
        self.device_name = kwargs["deviceName"]
        self.device_type = kwargs["deviceType"]
        self.hub_device_id = kwargs["hubDeviceId"]


@dataclass
class Remote:
    """Remote device."""

    device_id: str
    device_name: str
    device_type: str
    hub_device_id: str

    def __init__(self, **kwargs) -> None:
        """Initialize."""
        self.device_id = kwargs["deviceId"]
        self.device_name = kwargs["deviceName"]
        self.device_type = kwargs["remoteType"]
        self.hub_device_id = kwargs["hubDeviceId"]


class PowerState(Enum):
    """Power state."""

    ON = "on"
    OFF = "off"


class Commands(Enum):
    pass


class CommonCommands(Commands):
    """Common commands."""

    ON = "turnOn"
    OFF = "turnOff"


class OthersCommands(Commands):
    """Others commands."""

    CUSTOMIZE = "customize"  # Command {user-defined button name}


class AirConditionerCommands(Commands):
    """Air conditioner commands."""

    SET_ALL = "setAll"  # parameter: {temperature},{mode},{fan speed},{power state}


class HumidifierCommands(Commands):
    """Humidifier commands."""

    SET_MODE = "setMode"  # parameter: auto, set to Auto Mode, 101, set atomization efficiency to 34%,102, set atomization efficiency to 67%, 103, set atomization efficiency to 100%


class TVCommands(Commands):
    """TV commands."""

    SET_CHANNEL = "SetChannel"
    VOLUME_ADD = "volumeAdd"
    VOLUME_SUB = "volumeSub"
    CHANNEL_ADD = "channelAdd"
    CHANNEL_SUB = "channelSub"


class DVDCommands(Commands):
    """DVD commands."""

    SET_MUTE = "setMute"
    FAST_FORWARD = "FastForward"
    REWIND = "Rewind"
    NEXT = "Next"
    PREVIOUS = "Previous"
    PAUSE = "Pause"
    PLAY = "Play"
    STOP = "Stop"


class SpeakerCommands(Commands):
    """Speaker commands."""

    VOLUME_ADD = "volumeAdd"
    VOLUME_SUB = "volumeSub"


class FanCommands(Commands):
    """Fan commands."""

    SWING = "swing"
    TIMER = "timer"
    LOW_SPEED = "lowSpeed"
    MIDDLE_SPEED = "middleSpeed"
    HIGH_SPEED = "highSpeed"


class LightCommands(Commands):
    """Light commands."""

    BRIGHTNESS_UP = "brightnessUp"
    BRIGHTNESS_DOWN = "brightnessDown"


T = TypeVar("T", bound=CommonCommands)


class SwitchBotAPI:
    """SwitchBot API."""

    def __init__(self, token: str, secret: str) -> None:
        """Initialize."""
        self.token = token
        self.secret = secret

    def make_headers(self, token: str, secret: str):
        """Make headers."""
        nonce = uuid.uuid4()
        timestamp = int(round(time.time() * 1000))
        string_to_sign = bytes(f"{token}{timestamp}{nonce}", "utf-8")
        secret_bytes = bytes(secret, "utf-8")

        sign = base64.b64encode(hmac.new(secret_bytes, msg=string_to_sign, digestmod=hashlib.sha256).digest())

        return {
            "Authorization": token,
            "Content-Type": "application/json",
            "charset": "utf8",
            "t": str(timestamp),
            "sign": str(sign, "utf-8"),
            "nonce": str(nonce),
        }

    async def _request_device(self, path: str = "", callback: str = "get", json=None):
        async with ClientSession() as session:
            async with getattr(session, callback)(
                f"{_API_HOST}/v1.1/devices/{path}",
                headers=self.make_headers(self.token, self.secret),
                json=json,
            ) as response:
                if response.status == 403:
                    raise InvalidAuth()
                body = await response.json()
                if response.status >= 400 or body.get("statusCode") != 100:
                    _LOGGER.error("Error %s: %s", response.status, body)
                    raise CannotConnect()
                return body.get("body")

    async def list_devices(self):
        """List devices."""
        body = await self._request_device("")
        _LOGGER.debug("Devices: %s", body)
        devices = [Device(**device) for device in body.get("deviceList")]
        remotes = [
            Remote(**remote)
            for remote in body.get("infraredRemoteList")
            if remote.get("remoteType") not in NON_OBSERVED_REMOTE_TYPES
        ]
        return [*devices, *remotes]

    async def get_status(self, device_id: str):
        """No status for IR devices."""
        body = await self._request_device(f"{device_id}/status")
        return body

    async def send_command(
        self,
        device_id: str,
        command: T,
        command_type: str = "command",
        parameters: dict | str = "default",
    ):
        """Send command to device.

        Args:
            device_id (str): The ID of the device.
            command (extends CommonCommands): The command to be sent.
            command_type (str, optional): The type of the command. Defaults to "command".
            parameters (dict | str, optional): The parameters for the command. Defaults to "default".

        Example JSON:
            {
                "commandType": "customize",
                "command": "ボタン", // the name of the customized button
                "parameter": "default"
            }
        """
        json = {
            "commandType": command_type,
            "command": command.value,
            "parameter": parameters,
        }
        await self._request_device(f"{device_id}/commands", callback="post", json=json)
