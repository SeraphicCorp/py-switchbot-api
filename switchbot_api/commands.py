"""Base Commands for SwitchBot API."""

from __future__ import annotations

from enum import Enum
from typing import TypeVar


class Commands(Enum):
    """Base command class."""

    @classmethod
    def is_supported(cls, device_type: str) -> bool:
        """Is this commands supported this device type."""
        return device_type in cls.get_supported_devices()

    @classmethod
    def get_supported_devices(cls) -> list[str]:
        """Get supported devices."""
        return []


class CommonCommands(Commands):
    """Common commands."""

    ON = "turnOn"
    OFF = "turnOff"
    TOGGLE = "toggle"
    PRESS = "press"

    # Considering the wide range of CommonCommands, they are not inherited here.
    @classmethod
    def is_supported(cls, device_type: str) -> bool:
        """Is this commands supported this device type."""
        error_msg = "CommonCommands not implement this method"
        raise NotImplementedError(error_msg)

    @classmethod
    def get_supported_devices(cls) -> list[str]:
        """Get supported devices."""
        error_msg = "CommonCommands not implement this method"
        raise NotImplementedError(error_msg)


class BotCommands(Commands):
    """Bot commands."""

    PRESS = "press"

    @classmethod
    def get_supported_devices(cls) -> list[str]:
        """Get supported devices."""
        return ["Bot"]


class OthersCommands(Commands):
    """Others commands."""

    CUSTOMIZE = "customize"  # Command {user-defined button name}


class CurtainCommands(Commands):
    """Curtain & Curtain3 commands."""

    SET_POSITION = "setPosition"  # parameter(str): index0,mode0,position0  e.g. 0,ff,80
    PAUSE = "pause"

    @classmethod
    def get_supported_devices(cls) -> list[str]:
        """Get supported devices."""
        return ["Curtain", "Curtain 3"]


class LockCommands(Commands):
    """Lock commands."""

    LOCK = "lock"
    UNLOCK = "unlock"

    @classmethod
    def get_supported_devices(cls) -> list[str]:
        """Get supported devices."""
        return ["Smart Lock", "Smart Lock Lite", "Smart Lock Pro", "Smart Lock Ultra"]


class LockV2Commands(Commands):
    """Lock commands."""

    LOCK = "lock"
    UNLOCK = "unlock"
    DEADBOLT = "deadbolt"

    @classmethod
    def get_supported_devices(cls) -> list[str]:
        """Get supported devices."""
        return ["Smart Lock", "Smart Lock Pro", "Smart Lock Ultra"]


class HumidifierCommands(Commands):
    """Humidifier commands."""

    # parameter: auto, set to Auto Mode
    # 101, set atomization efficiency to 34%
    # 102, set atomization efficiency to 67%
    # 103, set atomization efficiency to 100%
    SET_MODE = "setMode"

    @classmethod
    def get_supported_devices(cls) -> list[str]:
        """Get supported devices."""
        return ["Humidifier"]


class HumidifierV2Commands(Commands):
    """Humidifier 2 commands."""

    SET_MODE = "setMode"
    SET_CHILD_LOCK = "setChildLock"

    @classmethod
    def get_supported_devices(cls) -> list[str]:
        """Get supported devices."""
        return ["Humidifier 2"]


class AirPurifierCommands(Commands):
    """Purifier commands."""

    # Supported Device list:
    # Air Purifier VOC
    # Air Purifier Table VOC
    # Air Purifier PM2.5
    # Air Purifier Table PM2.5

    SET_MODE = "setMode"
    SET_CHILD_LOCK = "setChildLock"

    @classmethod
    def get_supported_devices(cls) -> list[str]:
        """Get supported devices."""
        return [
            "Air Purifier VOC",
            "Air Purifier Table VOC",
            "Air Purifier PM2.5",
            "Air Purifier Table PM2.5",
        ]


class AirConditionerCommands(Commands):
    """Air conditioner commands."""

    SET_ALL = "setAll"

    @classmethod
    def get_supported_devices(cls) -> list[str]:
        """Get supported devices."""
        return ["Air Conditioner"]


class SwitchCommands(Commands):
    """Switch commands."""

    # Supported Device list:
    # Relay Switch 1
    # Relay Switch 1PM
    SET_MODE = "setMode"

    @classmethod
    def get_supported_devices(cls) -> list[str]:
        """Get supported devices."""
        return ["Relay Switch 1", "Relay Switch 1PM"]


class Switch2PMCommands(Commands):
    """Switch commands."""

    # Supported Device list:
    # Relay Switch 2PM
    SET_MODE = "setMode"
    SET_POSITION = "setPosition"

    @classmethod
    def get_supported_devices(cls) -> list[str]:
        """Get supported devices."""
        return ["Relay Switch 2PM"]


class RGBWLightCommands(Commands):
    """RGBWLight commands."""

    # Supported Device list:
    # Strip Light
    SET_BRIGHTNESS = "setBrightness"
    SET_COLOR = "setColor"

    @classmethod
    def get_supported_devices(cls) -> list[str]:
        """Get supported devices."""
        return ["Strip Light"]


class RGBWWLightCommands(Commands):
    """RGBWwLight commands."""

    # Supported Device list:
    # Floor Lamp
    # Strip Light 3
    # Color Bulb
    SET_BRIGHTNESS = "setBrightness"
    SET_COLOR = "setColor"
    SET_COLOR_TEMPERATURE = "setColorTemperature"

    @classmethod
    def get_supported_devices(cls) -> list[str]:
        """Get supported devices."""
        return ["Strip Light 3", "Floor Lamp", "Color Bulb"]


class DoorBellCommands(Commands):
    """Door Bell commands."""

    ENABLE = "enableMotionDetection"
    DISABLE = "disableMotionDetection"

    @classmethod
    def get_supported_devices(cls) -> list[str]:
        """Get supported devices."""
        return ["Video Doorbell"]


class VacuumCommands(Commands):
    """Vacuum commands."""

    START = "start"
    STOP = "stop"
    DOCK = "dock"
    POW_LEVEL = "PowLevel"

    @classmethod
    def get_supported_devices(cls) -> list[str]:
        """Get supported devices."""
        return [
            "K10+",
            "K10+ Pro",
            "Robot Vacuum Cleaner S1",
            "Robot Vacuum Cleaner S1 Plus",
        ]


class VacuumCleanerV2Commands(Commands):
    """VacuumCleanerV2 commands."""

    # Supported Device list:
    # K20+ Pro
    # Robot Vacuum Cleaner K10+ Pro Combo
    START_CLEAN = "startClean"
    PAUSE = "pause"
    DOCK = "dock"
    SET_VOLUME = "setVolume"
    CHANGE_PARAM = "changeParam"

    @classmethod
    def get_supported_devices(cls) -> list[str]:
        """Get supported devices."""
        return ["K20+ Pro", "Robot Vacuum Cleaner K10+ Pro Combo"]


class VacuumCleanerV3Commands(Commands):
    """VacuumCleanerV3 commands."""

    # Supported Device list:
    # Floor Cleaning Robot S10
    # S20
    START_CLEAN = "startClean"
    PAUSE = "pause"
    DOCK = "dock"
    SET_VOLUME = "setVolume"
    CHANGE_PARAM = "changeParam"
    ADD_WATER_FOR_HUMI = "addWaterForHumi"
    SELF_CLEAN = "selfClean"

    @classmethod
    def get_supported_devices(cls) -> list[str]:
        """Get supported devices."""
        return ["Robot Vacuum Cleaner S10", "S20"]


class CeilingLightCommands(Commands):
    """Ceiling light commands."""

    # 1-100
    SET_BRIGHTNESS = "setBrightness"
    # 2700-6500
    SET_COLOR_TEMPERATURE = "setColorTemperature"

    @classmethod
    def get_supported_devices(cls) -> list[str]:
        """Get supported devices."""
        return ["Ceiling Light", "Ceiling Light Pro"]


class BlindTiltCommands(Commands):
    """Blind Tilt commands."""

    SET_POSITION = "setPosition"
    FULLY_OPEN = "fullyOpen"
    CLOSE_UP = "closeUp"
    CLOSE_DOWN = "closeDown"

    @classmethod
    def get_supported_devices(cls) -> list[str]:
        """Get supported devices."""
        return ["Blind Tilt"]


class RollerShadeCommands(Commands):
    """Roller Shade commands."""

    SET_POSITION = "setPosition"

    @classmethod
    def get_supported_devices(cls) -> list[str]:
        """Get supported devices."""
        return ["Roller Shade"]


class BatteryCirculatorFanCommands(Commands):
    """Command types for [Battery Circulator Fan] API."""

    SET_WIND_SPEED = "setWindSpeed"
    SET_WIND_MODE = "setWindMode"
    SET_NIGHT_LIGHT_MODE = "setNightLightMode"

    @classmethod
    def get_supported_devices(cls) -> list[str]:
        """Get supported devices."""
        return ["Circulator Fan", "Battery Circulator Fan"]


class TVCommands(Commands):
    """TV commands."""

    SET_CHANNEL = "SetChannel"
    VOLUME_ADD = "volumeAdd"
    VOLUME_SUB = "volumeSub"
    CHANNEL_ADD = "channelAdd"
    CHANNEL_SUB = "channelSub"


class DVDCommands(Commands):
    """DVD commands."""

    SET_MUTE = "setMute"
    FAST_FORWARD = "FastForward"
    REWIND = "Rewind"
    NEXT = "Next"
    PREVIOUS = "Previous"
    PAUSE = "Pause"
    PLAY = "Play"
    STOP = "Stop"


class SpeakerCommands(Commands):
    """Speaker commands."""

    VOLUME_ADD = "volumeAdd"
    VOLUME_SUB = "volumeSub"


class FanCommands(Commands):
    """Fan commands."""

    SWING = "swing"
    TIMER = "timer"
    LOW_SPEED = "lowSpeed"
    MIDDLE_SPEED = "middleSpeed"
    HIGH_SPEED = "highSpeed"


class LightCommands(Commands):
    """Light commands."""

    BRIGHTNESS_UP = "brightnessUp"
    BRIGHTNESS_DOWN = "brightnessDown"


T = TypeVar("T", bound=CommonCommands)
