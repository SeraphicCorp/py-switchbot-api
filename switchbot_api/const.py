"""Constants for SwitchBot API."""

from enum import StrEnum


class Model(StrEnum):
    """Model of SwitchBot device."""

    HUB_2 = "Hub 2"
    CURTAIN = "Curtain"
    CURTAIN3 = "Curtain3"
