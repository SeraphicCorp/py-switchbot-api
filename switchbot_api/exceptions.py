"""Exceptions for the SwitchBot API."""


class SwitchBotError(Exception):
    """General SwitchBot error occurred."""


class SwitchBotConnectionError(SwitchBotError):
    """Cannot connect to the SwitchBot API."""


class SwitchBotAuthenticationError(SwitchBotError):
    """Invalid auth for the SwitchBot API."""


class SwitchBotDeviceOfflineError(SwitchBotError):
    """Device currently offline."""
