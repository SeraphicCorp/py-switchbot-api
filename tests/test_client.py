"""Tests for the client."""

from __future__ import annotations

import asyncio
from typing import TYPE_CHECKING, Any

import aiohttp
from aiohttp import ClientError
from aiohttp.hdrs import METH_GET, METH_POST
from aioresponses import CallbackResult, aioresponses
import pytest

from switchbot_api import SwitchBotAPI
from tests import load_fixture
from tests.const import MOCK_URL

if TYPE_CHECKING:
    from syrupy import SnapshotAssertion



async def test_device_list(
    responses: aioresponses,
    client: SwitchBotAPI,
    snapshot: SnapshotAssertion,
) -> None:
    """Test searching for media."""
    responses.get(
        f"{MOCK_URL}/devices",
        status=200,
        body=load_fixture("device_list.json"),
    )
    assert await client.list_devices() == snapshot
    # responses.assert_called_once_with(
    #     f"{MOCK_URL}/search",
    #     METH_GET,
    #     headers=HEADERS,
    #     params={"query": "frosty"},
    #     json=None,
    # )