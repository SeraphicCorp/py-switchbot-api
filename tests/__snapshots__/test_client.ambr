# serializer version: 1
# name: test_device_list
  list([
    dict({
      'device_id': 'D82812821F08',
      'device_name': 'Hub 2 08',
      'device_type': 'Hub 2',
      'hub_device_id': '',
    }),
    dict({
      'device_id': 'E2562C1DCC1A',
      'device_name': 'Curtain 1A',
      'device_type': 'Curtain3',
      'hub_device_id': 'D82812821F08',
    }),
    dict({
      'device_id': 'FCE84BB8B04F',
      'device_name': 'Curtain 4F',
      'device_type': 'Curtain',
      'hub_device_id': '',
    }),
    dict({
      'device_id': 'FF711F51601E',
      'device_name': 'Curtain 1E',
      'device_type': 'Curtain3',
      'hub_device_id': 'D82812821F08',
    }),
    dict({
      'device_id': '01-202501201703-63754985',
      'device_name': 'TV',
      'device_type': 'DIY TV',
      'hub_device_id': 'D82812821F08',
    }),
  ])
# ---
# name: test_device_status[curtain3]
  dict({
    'battery': 44,
    'calibrate': True,
    'deviceId': 'FF711F51601E',
    'deviceType': 'Curtain3',
    'group': False,
    'hubDeviceId': 'D82812821F08',
    'moving': False,
    'slidePosition': 0,
    'version': 'V1.2',
  })
# ---
# name: test_device_status[curtain]
  dict({
    'battery': 0,
    'calibrate': False,
    'deviceId': 'FCE84BB8B04F',
    'deviceType': 'Curtain',
    'group': False,
    'hubDeviceId': '000000000000',
    'moving': False,
    'slidePosition': 0,
    'version': 'V6.3',
  })
# ---
# name: test_device_status[hub_2]
  dict({
    'deviceId': 'D82812821F08',
    'deviceType': 'Hub 2',
    'hubDeviceId': 'D82812821F08',
    'humidity': 52,
    'lightLevel': 1,
    'temperature': 17.8,
    'version': 'V2.3-1.4',
  })
# ---
